### ISSUE TITLE
Input Field Labels

### ISSUE THREAD

**Issue by blandfried** [okta/okta-signin-widget#356]
Not sure if I'm just missing it in the docs, but can the option to have labels for the input fields be added? Right now there are no labels and just help text and placeholder text. This would help with accessibility and Section 508 guidelines. I was able to create the image below using some JavaScript selectors but I think it should be a built-in option for the widget.

![sign-in-widget-with-labels](https://user-images.githubusercontent.com/1953556/33340306-ca6ec612-d449-11e7-81a1-d58d038f1317.PNG)

**Comment by santhoshbalakrishnan** [comment]
As far as I know there is no option to add a label using sign-in widget options. But, it does have aria-labels on the input fields. For example, `aria-label="Username"` on the username field and that should help with accessibility.

**Comment by blandfried** [comment]
Section 508 is just written vague enough that it's hard to know what they want sometimes. I guess it seems like the combination of aria-label, placeholder and tool tip is enough to make it pass 508 validation. I still wouldn't be against a label option being added to the widget.

[https://www.section508.va.gov/support/webdev/4_2.asp](url)
[https://www.section508.va.gov/support/webdev/4_5b.asp](url)

**Comment by hor-kanchan-okta** [comment]
The aria-label is added for accessibility. section 508 1194.22 n
https://github.com/okta/okta-signin-widget/pull/129

**Comment by rchild-okta** [comment]
Let's mark this as a feature request - i.e. the option to shift placeholders to labels that sit outside the form inputs. @blandfried not sure when we'd be able to get to it, but we can provide this feedback to our design team and see what they think.

**Comment by maggielaw-okta** [comment]
This WebAIM article provides helpful guidance: https://webaim.org/techniques/forms/advanced

See the "Invisible Labels" section toward the bottom. Echoing @hor-kanchan-okta and @santhoshbalakrishnan have noted, the `aria-label` attribute does the trick. It's read by screen readers so makes this compliant. 

For sighted users, a `placeholder` serving as the sole visible label makes sense in certain cases. You wouldn't want to do that everyone, but for a login form, the conventions are strong and pattern recognizable. Explicit labels are arguably overkill here, in fact. 

The article warns against combinations that create redundant readouts by the AT -- e.g., a `title` and a hidden `<label>` -- but they're fine with `placeholder` and `aria-label`.

**Comment by hor-kanchan-okta** [comment]
Since the design team chose not to include the label and our current approach is 508 compliance and WCAG 2.0, I am closing this issue.

**Comment by sawyerh** [comment]
> the conventions are strong and pattern recognizable. Explicit labels are arguably overkill here, in fact. Explicit labels are arguably overkill here

@hor-kanchan-okta @maggielaw-okta @rchild-okta I'd urge the Okta design team to reconsider this position of not supporting `<label>` elements, especially since placeholders are also the primary labels on the Registration page, where a number of custom fields can be displayed that don't fit a "strong and recognizable pattern."

You've shared links pointing out how the current implementation is compliant, but compliance and usability are two very different things. Through a bounty of usability testing from groups like GOV.UK, Nielsen Norman Group, larger companies, and accessibility advocates since the original closing of this issue, this pattern has proved to _not_ be the most friendly for all user groups:

- [Nielsen Norman Group: Placeholders in Form Fields Are Harmful](https://www.nngroup.com/articles/form-design-placeholders/)
- [W3C Placeholder Research](https://www.w3.org/WAI/GL/low-vision-a11y-tf/wiki/Placeholder_Research)
- [GOV.UK requires form labels](https://design-system.service.gov.uk/components/text-input/#label-text-inputs)
- [U.S. Web Design System discourages placeholder text](https://designsystem.digital.gov/components/form-controls/)

**Comment by maggielaw-okta** [comment]
@sawyerh gosh that was a very old comment. We're definitely moving into a direction of explicit form labels, with SIW and elsewhere (https://design-docs.trexcloud.com/components/form.htm). It will take a while but it's the right thing to do for a11y. Thank you for your input!

Cc: @edburyenegren-okta

**Comment by sawyerh** [comment]
@maggielaw-okta @edburyenegren-okta Awesome, that's great to hear, thank you!

**Comment by william-holt** [comment]
@maggielaw-okta @edburyenegren-okta I was wondering if any progress had been made on adding explicit form labels to the widget options? I can't seem to find any mention in the docs and just stumbled across this old issue.

**Comment by edburyenegren-okta** [comment]
Hi @william-holt, as I understand it this has been addressed in the most recent iteration of SIW, as seen on our homepage: https://developer.okta.com/
